### ISSUE TITLE
[FEA] JIT Compilation of CUDF Kernels

### ISSUE THREAD

**Issue by lamarrr** [rapidsai/cudf#17399]
**Is your feature request related to a problem? Please describe.**
As described in #15366, we intend to adopt JIT compilation to some of our kernels using JITify/NVRTC.
JITify presently only supports compilation of device code, this means we can't mix host code with it, like we'd do with NVCC. Some important headers are not supported, i.e.
- <stdexcept>
- <atomic> (replace with cuda/std/atomic)

**Describe the solution you'd like**
We'd need to:
- Separate our kernel headers into device-only code headers
- Patch Thrust and RMM headers to separate their headers into device-only code headers

**Describe alternatives you've considered**
- Using macros to separate the device-only code from host code, this doesn't work and would have been very brittle
- Depending on NVCC for PTX and loading it into the driver at runtime

**Comment by jrhemstad** [comment]
> Patch Thrust and RMM headers to separate their headers into device-only code headers

Please no. RAPIDS is banned from applying any more patches to Thrust headers 🙂 

You should only include RMM or Thrust headers in the host-only headers, so there shouldn't be any need to modify those.

**Comment by jrhemstad** [comment]
> > Please no. RAPIDS is banned from applying any more patches to Thrust headers 🙂
> 
> Even locally within cudf?

Yes. Local patches to CCCL code causes all sorts of problems as the patch files need to be updated anytime any line of code in the patched file is updated. CCCL runs RAPIDS in its CI and patches become a nightmare.

**Comment by vyasr** [comment]
For CCCL in particular we are aiming to get all patches that we have historically needed upstreamed so that we can rely on CCCL's CI like Jake mentioned. Also more generally we do not want to ship patched libraries any more. It causes loads of unrelated potential packaging problems down the line.
